module/zfs/spa_errlog.c

```diff
@@ -128,6 +130,103 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
 	mutex_exit(&spa->spa_errlist_lock);
 }
 
+/*
+ * If this error exists in the given tree remove it.
+ */
+static void
+remove_error_from_list(spa_t *spa, avl_tree_t *t, const zbookmark_phys_t *zb)
+{
+	spa_error_entry_t search, *found;
+	avl_index_t where;
+
+	mutex_enter(&spa->spa_errlist_lock);
+	search.se_bookmark = *zb;
+	if ((found = avl_find(t, &search, &where)) != NULL)
+		avl_remove(t, found);
+	mutex_exit(&spa->spa_errlist_lock);
+}
+
+/*
+ * If a healed bookmark matches an entry in the error log we stash it in a tree
+ * so that we can later remove the related log entries in sync context.
```

Comment:
Could there be a race condition where the entry is in the in-memory errlist but not the on-disk errlog, so we don't clear it?  It seems like it would be simpler for this function to always add an entry to the spa_errlist_healed (if it isn't already there), and let spa_remove_healed_errors() handle the case where it isn't actually in the on-disk structure (as it already does).